server/src/main/java/com/linecorp/centraldogma/server/internal/api/GitHttpService.java

```diff
@@ -0,0 +1,206 @@
+                                   "Unsupported service: " + service);
+        }
+        repoName = maybeRemoveGitSuffix(repoName);
+        if (gitProtocol == null || !gitProtocol.contains(VERSION_2_REQUEST)) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported git-protocol: " + gitProtocol);
+        }
+
+        if (!projectApiManager.exists(projectName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Project not found: " + projectName);
+        }
+        if (!projectApiManager.getProject(projectName).repos().exists(repoName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Repository not found: " + repoName);
+        }
+        return CAPABILITY_ADVERTISEMENT_RESPONSE.toHttpResponse();
+    }
+
+    private static String maybeRemoveGitSuffix(String repoName) {
+        if (repoName.length() > 5 && repoName.endsWith(".git")) {
+            repoName = repoName.substring(0, repoName.length() - 4);
+        }
+        return repoName;
+    }
+
+    // https://www.git-scm.com/docs/gitprotocol-http#_smart_service_git_upload_pack
+    @Post("/{projectName}/{repoName}/git-upload-pack")
+    public HttpResponse gitUploadPack(AggregatedHttpRequest req,
+                                      @Param String projectName, @Param String repoName) {
+        repoName = maybeRemoveGitSuffix(repoName);
+        final String gitProtocol = req.headers().get("git-protocol");
+        if (gitProtocol == null || !gitProtocol.contains(VERSION_2_REQUEST)) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported git-protocol: " + gitProtocol);
+        }
+
+        final MediaType contentType = req.headers().contentType();
+        if (contentType == null || !"application/x-git-upload-pack-request".equals(contentType.toString())) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported content-type: " + contentType);
+        }
+
+        if (!projectApiManager.exists(projectName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Project not found: " + projectName);
+        }
+        if (!projectApiManager.getProject(projectName).repos().exists(repoName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Repository not found: " + repoName);
+        }
+
+        final Repository jGitRepository =
+                projectApiManager.getProject(projectName).repos().get(repoName).jGitRepository();
+
+        final ByteStreamMessage body = StreamMessage.fromOutputStream(os -> {
```

Comment:
You mean just adding `@Blocking` to this method or is there something else?

Reply:
I thought of adding `@Blocking` to the service method.

Reply:
Had a chat with @minwoox and I found out that he used `StreamMessage.fromOutputStream()` to support backpressure in case the returned git objects are huge.

I agree with his opinion and using `StreamMessage.fromOutputStream()` makes more sense.